server/src/main/java/org/zanata/mt/process/DocumentProcessManager.java

```diff
@@ -25,31 +31,38 @@
     private static final Logger LOG =
             LoggerFactory.getLogger(DocumentProcessManager.class);
 
-    private final ReentrantLock lock = new ReentrantLock(true);
+    private final StripedLock lock = new StripedLock();
 
     @SuppressWarnings("unused")
     public DocumentProcessManager() {
     }
 
-    public void lock(@NotNull DocumentProcessKey key) {
+    private void lock(@NotNull DocumentProcessKey key) {
         if (LOG.isDebugEnabled()) {
             LOG.debug("Locking document translation request:{}", key.toString());
         }
-        lock.lock();
+        lock.acquireLock(key, true);
     }
 
-    public void unlock(@NotNull DocumentProcessKey key) {
+    private void unlock(@NotNull DocumentProcessKey key) {
         if (LOG.isDebugEnabled()) {
             LOG.debug("release document translation lock:{}", key.toString());
         }
-        lock.unlock();
+        lock.releaseLock(key);
     }
 
-    public boolean isLocked(@NotNull DocumentProcessKey key) {
-        return lock.isLocked();
+    public Response withLock(DocumentProcessKey key, Callable<Response> callable)
+            throws Exception {
```

Comment:
Ah. Rather than polluting all the method signatures with `throws Exception`, we should probably use `java.util.function.Function`.